File: mc/rcmd/src/main/java/calebxzhou/rdi/mc/rcmd/RcmdContext.java
package calebxzhou.rdi.mc.rcmd;

import java.util.List;
import java.util.Map;

public record RcmdContext(
        RcmdSource source,
        String rawInput,
        RcmdCommandSpec spec,
        List<String> tokens,
        Map<String, Object> arguments
) {
    public RcmdContext {
        tokens = List.copyOf(tokens);
        arguments = Map.copyOf(arguments);
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String name) {
        return (T) arguments.get(name);
    }

    public String getString(String name) {
        return get(name);
    }

    public boolean getBool(String name) {
        return get(name);
    }

    public int getInt(String name) {
        return get(name);
    }

    public long getLong(String name) {
        return get(name);
    }

    public double getDouble(String name) {
        return get(name);
    }
}

File: mc/rcmd/src/main/kotlin/calebxzhou/rdi/mc/rcmd/RcmdContext.kt
package calebxzhou.rdi.mc.rcmd

class RcmdContext(
    val source: RcmdSource,
    val rawInput: String,
    val spec: RcmdCommandSpec,
    tokens: List<String>,
    arguments: Map<String, Any>
) {
    val tokens: List<String> = tokens.toList()
    val arguments: Map<String, Any> = arguments.toMap()

    @Suppress("UNCHECKED_CAST")
    fun <T> get(name: String): T = arguments[name] as T

    fun getString(name: String): String = get(name)

    fun getBool(name: String): Boolean = get(name)

    fun getInt(name: String): Int = get(name)

    fun getLong(name: String): Long = get(name)

    fun getDouble(name: String): Double = get(name)
}
